Cython/Compiler/Optimize.py

```diff
@@ -2773,6 +2775,31 @@ def _handle_simple_function_len(self, node, function, pos_args):
             new_node = ExprNodes.PythonCapiCallNode(
                 node.pos, "__Pyx_MemoryView_Len", func_type,
                 args=[arg], is_temp=node.is_temp)
+        elif arg.type.is_unicode_char:
+            new_node = ExprNodes.IntNode(node.pos, value='1', constant_result=1,
+                                     type=node.type)
+        elif arg.type.is_ctuple:
+            new_node = ExprNodes.IntNode(node.pos, value=str(arg.type.size), constant_result=1,
+                                     type=node.type)
+
+            # TODO: This might fail if the source of the ctuple has side effects and isn't a sequence constructor
+            if isinstance(arg, ExprNodes.TypecastNode) and arg.operand.is_sequence_constructor:
+                for arg2 in arg.operand.args:
+                    if isinstance(arg2, ExprNodes.CoerceToPyTypeNode):
+                        arg2 = arg2.arg
+
+                    if not arg2.try_is_simple():
+                        new_node = UtilNodes.EvalWithTempExprNode(UtilNodes.LetRefNode(arg2), new_node)
+        elif arg.is_sequence_constructor and not arg.is_starred:
+            new_node = ExprNodes.IntNode(node.pos, value=str(len(arg.args) * arg.mult_factor), constant_result=1,
+                                     type=node.type)
+
+            for arg2 in arg.args:
+                if isinstance(arg2, ExprNodes.CoerceToPyTypeNode):
+                    arg2 = arg2.arg
+
+                if not arg2.try_is_simple():
+                    new_node = UtilNodes.EvalWithTempExprNode(UtilNodes.LetRefNode(arg2), new_node)
```

Comment:
I think the same here - emulating the side effects doesn't optimize much and is pretty complicated.